components/application-operator/pkg/kymahelm/gateway/gatewaymanager.go

```diff
@@ -0,0 +1,199 @@
+	"github.com/kyma-project/kyma/components/application-operator/pkg/kymahelm"
+	"github.com/kyma-project/kyma/components/application-operator/pkg/utils"
+	"github.com/pkg/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	gatewayChartDirectory = "gateway"
+	gatewayNameFormat     = "%s-application-gateway"
+)
+
+//go:generate mockery -name GatewayManager
+type GatewayManager interface {
+	InstallGateway(namespace string) error
+	DeleteGateway(namespace string) error
+	GatewayExists(namespace string) (bool, error)
+	UpgradeGateways() error
+}
+
+//go:generate mockery -name ServiceCatalogueClient
+type ServiceCatalogueClient interface {
+	ServiceInstances(namespace string) v1beta1.ServiceInstanceInterface
+}
+
+func NewGatewayManager(helmClient kymahelm.HelmClient, overrides OverridesData, serviceCatalogueClient ServiceCatalogueClient, namespaceClient v12.NamespaceInterface) GatewayManager {
+	return &gatewayManager{
+		helmClient:             helmClient,
+		overrides:              overrides,
+		serviceCatalogueClient: serviceCatalogueClient,
+		namespaces:             namespaceClient,
+	}
+}
+
+type gatewayManager struct {
+	helmClient             kymahelm.HelmClient
+	overrides              OverridesData
+	serviceCatalogueClient ServiceCatalogueClient
+	namespaces             v12.NamespaceInterface
+}
+
+func (g *gatewayManager) InstallGateway(namespace string) error {
+	overrides, err := kymahelm.ParseOverrides(g.overrides, overridesTemplate)
+	if err != nil {
+		return errors.Errorf("Error parsing overrides: %s", err.Error())
+	}
+
+	name := getGatewayName(namespace)
+
+	_, err = g.helmClient.InstallReleaseFromChart(gatewayChartDirectory, namespace, name, overrides)
+	if err != nil {
+		return errors.Errorf("Error installing Gateway: %s", err.Error())
+	}
+	return nil
+}
+
+func (g *gatewayManager) DeleteGateway(namespace string) error {
+	gateway := getGatewayName(namespace)
+	releaseExist, _, err := g.gatewayExists(gateway, namespace)
+	if err != nil {
+		return errors.Errorf("Error deleting Gateway: %s", err.Error())
+	}
+	if releaseExist {
+		return g.deleteGateway(gateway)
+	}
+	return nil
+}
+
+func (g *gatewayManager) deleteGateway(gateway string) error {
+	_, err := g.helmClient.DeleteRelease(gateway)
+	if err != nil {
+		return errors.Errorf("Error deleting Gateway: %s", err.Error())
+	}
+
+	return nil
+}
+
+func (g *gatewayManager) GatewayExists(namespace string) (bool, error) {
+	name := getGatewayName(namespace)
+	exists, _, err := g.gatewayExists(name, namespace)
+	return exists, err
+}
+
+func (g *gatewayManager) UpgradeGateways() error {
+	namespaces, err := g.getAllNamespacesWithServiceInstances()
+
+	if err != nil {
+		return errors.Errorf("Error updating Gateway: %s", err.Error())
+	}
+
+	if len(namespaces) == 0 {
+		return nil
+	}
+
+	err = g.updateGateways(namespaces)
+
+	if err != nil {
+		return errors.Errorf("Error updating Gateway: %s", err.Error())
+	}
```

Comment:
If we for some reason cannot update the Gateway release, we end up in the crash loop and the whole Application Operator is not working. I do not think it is the desired behavior and we should just log and error and proceed as we do in case of Applications.